### PR TITLE
fixed: material out of range, boolean mod, new: gdscript, gdprim, gdcsg, grease pencil, reexport button

### PR DESCRIPTION
**Pull request by hartsantler**
for some reason an empty set ends up in the modifiers list.  i have not tested yet with many other modifiers.  i'm running the official blender2.8

there was also some issue with linked materials, i just have it fall back to the last material, but not sure if that is fully correct.

### CODE REVIEW

**Comment by Jason0214** [comment]
Could you separate your PR so that every one of them tracks individual issue, looks like there are three bug fixes in this PR. wow, thanks for your work！

For your information, you can create three fork of current master and use `git cherry-pick -i <commit-hash>` to cherry pick your commits onto those forks respectively.

File: io_scene_godot/__init__.py [hunk 1]
```diff
@@ -39,6 +39,83 @@
 }
 
 
+class GodotTextProps(bpy.types.Panel):
```
Comment (Jason0214) [inline io_scene_godot/__init__.py]: Could you move this into another pull request? thanks

Also probably better to put into another script something like `ui_components.py`, there will be other thing we can do to have more UI support in the future.

File: io_scene_godot/converters/animation/serializer.py [hunk 2]
```diff
@@ -88,6 +88,8 @@ def factory(cls, xform_matrix, rotation_mode='QUATERNION'):
 
         if rotation_mode == 'QUATERNION':
             xform_frame.rotation_euler = xform_matrix.to_euler()
+        elif rotation_mode == 'AXIS_ANGLE':
```
Comment (Jason0214) [inline io_scene_godot/converters/animation/serializer.py]: Could move this change into another pull request? so I can quickly merge it, thanks.

Also you can merge this two branch into `if rotation_mode in ('QUATERNION', 'AXIS_ANGLE')`

File: io_scene_godot/converters/grease.py [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+    if has_physics(obj):
+        mesh_node['transform'] = mathutils.Matrix.Identity(4)
+    else:
+        mesh_node['transform'] = obj.matrix_local
+
+    escn_file.add_node(mesh_node)
+    return mesh_node
+
+
+def fix_vertex(vtx):
```
Comment (Jason0214) [inline io_scene_godot/converters/grease.py]: Probably move this function into utils.py  and sharing with mesh.py

File: io_scene_godot/converters/grease.py [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+    """Changes a single position vector from y-up to z-up"""
+    return mathutils.Vector((vtx.x, vtx.z, -vtx.y))
+
+
+
+
+class ArrayMeshResource(InternalResource):
```
Comment (Jason0214) [inline io_scene_godot/converters/grease.py]: Same comment as above

File: io_scene_godot/converters/grease.py [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+                    #vertex_indices.append(vertex_index)
+
+                #if len(vertex_indices) > 2:  # Only triangles and above
+                #    surface.vertex_data.indices.append(vertex_indices)
+
+        for surface in surfaces:
+            self.mesh_resource[surface.name_str] = surface
+
+
+class VerticesArrays:
```
Comment (Jason0214) [inline io_scene_godot/converters/grease.py]: Not have much knowledge of Grease Object, so is these tagent, UV, etc all necessary?

File: io_scene_godot/converters/grease.py [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+        surface_lines.append(face_indices.to_string())
+
+        return surface_lines
+
+    def to_string(self):
+        """Serialize"""
+        return self.generate_lines().to_string()
+
+
+class Surface:
```
Comment (Jason0214) [inline io_scene_godot/converters/grease.py]: Probably put into `utils.py` and share with `mesh.py`

File: io_scene_godot/converters/grease.py [hunk 7]
```diff
@@ -0,0 +1,367 @@
...
+
+# PRIMITIVE_POINTS = 0 — Render array as points (one vertex equals one point).
+# PRIMITIVE_LINES = 1 — Render array as lines (every two vertices a line is created).
+# PRIMITIVE_LINE_STRIP = 2 — Render array as line strip.
+# PRIMITIVE_LINE_LOOP = 3 — Render array as line loop (like line strip, but closed).
+# PRIMITIVE_TRIANGLES = 4 — Render array as triangles (every three vertices a triangle is created).
+# PRIMITIVE_TRIANGLE_STRIP = 5 — Render array as triangle strips.
+# PRIMITIVE_TRIANGLE_FAN = 6 — Render array as triangle fans.
+
+class Vertex:
```
Comment (Jason0214) [inline io_scene_godot/converters/grease.py]: Same comment

File: io_scene_godot/converters/grease.py [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+        surface_object['primitive'] = 3  # simpler to debug
+        surface_object['arrays'] = self.vertex_data
+        surface_object['morph_arrays'] = self.morph_arrays
+        return surface_object
+
+    def to_string(self):
+        """Serialize"""
+        return self.generate_object().to_string()
+
+# PRIMITIVE_POINTS = 0 — Render array as points (one vertex equals one point).
```
Comment (Jason0214) [inline io_scene_godot/converters/grease.py]: Instead of making them all comments, maybe you can define it as global variable, it would be more readable when use them in the code.

e.g.
```
# PRIMITIVE_POINTS = 0 — Render array as points (one vertex equals one point).
```

Could be changed to 
```
PRIMITIVE_POINTS = 0  # Render array as points (one vertex equals one point).
```
And put them into `utils.py` so that they can be shared with `mesh.py`

File: io_scene_godot/converters/grease.py [hunk 9]
```diff
@@ -0,0 +1,367 @@
...
+
+    def generate_object(self):
+        """Generate object with mapping structure which fully
+        describe the surface"""
+        surface_object = Map()
+        if self.material is not None:
+            surface_object['material'] = self.material
+        #surface_object['primitive'] = 5  # triangle strip, sucks with complex shapes
+        #surface_object['primitive'] = 6  # triangle fan, looks better than strip
+        surface_object['primitive'] = 3  # simpler to debug
```
Comment (Jason0214) [inline io_scene_godot/converters/grease.py]: If you take my suggestion below, here it could be PRIMITIVE_LINE_LOOP

File: io_scene_godot/converters/material/material.py [hunk 10]
```diff
@@ -68,7 +68,11 @@ def generate_material_resource(escn_file, export_settings, bl_object,
         # to convert material to external file
         material_rsc_name = ''
 
-    if (engine in ('CYCLES', 'BLENDER_EEVEE') and
+    if material.is_grease_pencil:
+        mat = InternalResource("ShaderMaterial", material_rsc_name)
```
Comment (Jason0214) [inline io_scene_godot/converters/material/material.py]: Is suppose to be 'SpatialMaterial' ?

File: io_scene_godot/converters/mesh.py [hunk 11]
```diff
@@ -13,6 +13,99 @@
 
 MAX_BONE_PER_VERTEX = 4
 
+# ------------------------------- Prim -----------------------------------
```
Comment (Jason0214) [inline io_scene_godot/converters/mesh.py]: Could you move primitive and CSG support into another pull request? Let this pull request just for Grease. Separate PRs can be more effective to get things merged

File: io_scene_godot/export_godot.py [hunk 12]
```diff
@@ -121,9 +119,100 @@ def export_object(self, obj, parent_gd_node):
         # blender node exported as two parented node
         exported_node = exporter(self.escn_file, self.config, obj,
                                  parent_gd_node)
-
         self.bl_object_gd_node_map[obj] = exported_node
 
+        gds = None
```
Comment (Jason0214) [inline io_scene_godot/export_godot.py]: Move gdscript related change to another pull request, please.

Also I think it would be better to make another python file dedicated to gdscript related implementation. The main loop should be kept as small as possible

File: io_scene_godot/export_godot.py [hunk 13]
```diff
@@ -50,9 +50,7 @@ def find_godot_project_dir(export_path):
     while not os.path.isfile(os.path.join(project_dir, "project.godot")):
         project_dir = os.path.split(project_dir)[0]
         if project_dir in ("/", last):
```
Comment (Jason0214) [inline io_scene_godot/export_godot.py]: Could you explain why these changes?

File: io_scene_godot/structures.py [hunk 14]
```diff
@@ -93,10 +97,20 @@ def fix_paths(self, export_settings):
 
     def to_string(self):
         """Serializes the file ready to dump out to disk"""
+        ires = []
```
Comment (Jason0214) [inline io_scene_godot/structures.py]: I think if you always put `InternalResource()` into `self. internal_resources`, we would not need this change right?

File: io_scene_godot/structures.py [hunk 15]
```diff
@@ -453,6 +473,8 @@ def to_string(val):
         converter = CONVERSIONS.get(type(val))
         if converter is not None:
             val = converter(val)
+        elif val is None:
```
Comment (Jason0214) [inline io_scene_godot/structures.py]: If you really need this change, I think it better to put the `val is None` checking in front of all the other checkings

File: io_scene_godot/structures.py [hunk 16]
```diff
@@ -279,6 +296,9 @@ def add_elements(self, list_of_lists):
         for lis in list_of_lists:
             self.extend(lis)
 
+    def add_elements_flat(self, array):
```
Comment (Jason0214) [inline io_scene_godot/structures.py]: You can simple call `Array().extend()` to achieve this,  Array inherit list

Comment (hartsantler) [reply io_scene_godot/structures.py]: fixed

File: io_scene_godot/structures.py (hunk 15 above)
Comment (hartsantler) [reply io_scene_godot/structures.py]: fixed

File: io_scene_godot/structures.py (hunk 14 above)
Comment (hartsantler) [reply io_scene_godot/structures.py]: this change is helpful for others who are quickly jumping into this code base, adding some feature they need, and don't have much time to read how the exporter is designed.  By providing the option to pass a list or string, the exporter becomes easy to customize with minimal understanding of its internals.

File: io_scene_godot/export_godot.py (hunk 13 above)
Comment (hartsantler) [reply io_scene_godot/export_godot.py]: for testing i'm just exporting to /tmp, then i copy to my godot project folder later.
and later i want to have an option to generate the entire godot project folder.

File: io_scene_godot/converters/grease.py (hunk 5 above)
Comment (hartsantler) [reply io_scene_godot/converters/grease.py]: i'm also new to grease pencil, so i'm not sure if it has tangents or UV's

File: io_scene_godot/converters/grease.py (hunk 4 above)
Comment (hartsantler) [reply io_scene_godot/converters/grease.py]: fixed

File: io_scene_godot/converters/grease.py (hunk 3 above)
Comment (hartsantler) [reply io_scene_godot/converters/grease.py]: fixed

File: io_scene_godot/converters/animation/serializer.py (hunk 2 above)
Comment (hartsantler) [reply io_scene_godot/converters/animation/serializer.py]: fixed

File: io_scene_godot/__init__.py (hunk 1 above)
Comment (hartsantler) [reply io_scene_godot/__init__.py]: fixed

File: io_scene_godot/converters/grease.py (hunk 5 above)
Comment (hartsantler) [reply io_scene_godot/converters/grease.py]: i wish there was a way to detect blender prims, but there's no extra metadata on them.
one workaround is we provide our own buttons for making prims, which just call the blender prim ops, but we add the extra `gdprim` and `gdprim_radius` custom attributes automatically to the prim.

Comment (hartsantler) [reply io_scene_godot/converters/grease.py]: how the gd* attributes autogen a gdscript is complex, and probably better if i explain it a wiki, or other doc.  If you setup a wiki for this project, i will add a page there explaining how this works.

Comment (Jason0214) [reply io_scene_godot/converters/grease.py]: > i wish there was a way to detect blender prims, but there's no extra metadata on them.

Yeah, looks like there isn't any. Maybe use the custom attributes is the best thing we can do.

> how the gd* attributes autogen a gdscript is complex, and probably better if i explain it a wiki, or other doc. If you setup a wiki for this project, i will add a page there explaining how this works.

Create a wiki sounds very good. Unfortunately the wiki of github repo is not forkable, you can probably start writing the basic idea of your design in the wiki of your own [fork](https://github.com/hartsantler/godot-blender-exporter) just for the purpose of reviewing this PR ^ ^. 
And I can help to create a full document in [Godot Doc](https://docs.godotengine.org/en/3.1/getting_started/workflow/assets/escn_exporter/index.html?highlight=exporter) once the gdscript feature merged in.

File: io_scene_godot/export_godot.py (hunk 13 above)
Comment (Jason0214) [reply io_scene_godot/export_godot.py]: Okay, it's fine if you are doing it just for testing purpose, we can revert it before it merge in.

For your information, there is an [exporting option](https://github.com/godotengine/godot-blender-exporter/blob/master/io_scene_godot/__init__.py#L154) that can turn off exporting to a project dir.

File: io_scene_godot/structures.py (hunk 14 above)
Comment (Jason0214) [reply io_scene_godot/structures.py]: I understand your point, to make the code much flexible and context-free for new developers.  
However, I kind of disagree with this part of change, I also want the code to be manageable. For the InternalResource case, I think it is better just delegate all the serialization work to the InternalResource class. If you want insert a list of InternalResource objects, just add it to escn_file one by one.

Also the InternalResource is strictly corresponding to Resource in Godot, we are use it to detect multiple references to same Resource and avoid exporting one Resource more than once. Your change would break this functionality.

File: io_scene_godot/converters/utils.py [hunk 17]
```diff
@@ -1,7 +1,20 @@
 """Util functions and structs shared by multiple resource converters"""
+import bpy, bmesh, mathutils
 
-import bpy
-import bmesh
+gdenums = {
```
Comment (Jason0214) [inline io_scene_godot/converters/utils.py]: I think you can simply make `PRIMITIVE_POINTS` a global variable,  make a dict may have some unecessary performance issue.

Comment (hartsantler) [reply io_scene_godot/converters/utils.py]: you must be crazy bro. i don't want to read bullshit like this when i wake up.

File: io_scene_godot/converters/mesh.py [hunk 18]
```diff
@@ -278,7 +371,11 @@ def generate_surfaces(self, escn_file, export_settings, mesh):
                 surface.id = surface_index
                 surfaces.append(surface)
                 if mesh.materials:
-                    mat = mesh.materials[face.material_index]
+                    if face.material_index >= len(mesh.materials):
```
[the remark this answers is not in the document]
Comment (hartsantler) [reply io_scene_godot/converters/mesh.py]: its a complex scene, i do not know what causes it. not posting the blend file.

File: io_scene_godot/structures.py (hunk 14 above)
Comment (hartsantler) [reply io_scene_godot/structures.py]: my change breaks nothing.  Anyone who passes a list or string to that function, knows what they are doing, because they are adding some new generator, or inserting their own custom stuff, they understand and may prefer to manually manage the resource ids.

**Comment by hartsantler** [comment]
because i don't like your attitude, i'm closing this pull request and hard forking away.

you were born with that attitude, or is it something you learned since becoming yet another coder working at Google?

**Comment by Jason0214** [comment]
Are you referring to my attitude about having a somewhat standard for code quality?
Good for to notice, I am born with it and my Company emphasis it. If you don't like, you can just walk away. Fine with me.

**Comment by hartsantler** [comment]
Its obvious what attitude i'm talking about.  Its your Google Elitism, nobody likes it, and it forces us in California to enjoy hitting you and your cars with rocks.  Fuck you and everybody that works with you.
https://fortune.com/2018/01/18/apple-google-bus-attacks/

**Comment by akien-mga** [comment]
@hartsantler We don't accept such attitude from Godot Engine contributors, or anyone interacting with our users and developers.

@Jason0214 has been very patient and constructive in his comment, and if you don't want to comply with the quality standards of his project, that's fine, but you shouldn't insult him. You are free to walk away with your contributions unmerged if you don't want to do the extra work - others can pick it up from there.

I would ask that you edit your previous comments to remove the insults, otherwise I will have to ban you from commenting on Godot repositories.

**Comment by akien-mga** [comment]
And as mentioned previously, these changes should all be in separate PRs focusing on one issue each. @Jason0214 was quite patient reviewing this mega-PR, but it's not in any mergeable state.
